### PR TITLE
[FLINK-21229] add confluent schema registry ssl support

### PR DESCRIPTION
**Pull request by qingdao81**
## What is the purpose of the change

This pull requests adds ssl support to the confluent schema registry avro serializer/deserializer. 

## Brief change log

- `CachedSchemaCoderProvider` was extended to accept and delegate additional config properties to confluents `CachedSchemaRegistryClient`
- additional factory methods were added to `ConfluentRegistryAvroSerializationSchema` and `ConfluentRegistryAvroDeserializationSchema` which allow to pass a config map which is forwarded to `CachedSchemaCoderProvider`
- by definition all properties in this map with the prefix `schema.registry` are used to setup a `SslFactory` which is used in the http client for schema registry requests, see: [CachedSchemaRegistryClient](https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java#L180-L188)

## Verifying this change

This change added tests and can be verified as follows:

- Added unit tests for `ConfluentRegistryAvroSerializationSchema`, `ConfluentRegistryAvroDeserializationSchema` and `CachedSchemaCoderProvider`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by vidyaranya92** [comment]
Hi, Is there any expected timeline for releasing this?

**Comment by qingdao81** [comment]
I fixed the issues in the unit tests - now I'm waiting for someone to review this PR.

**Comment by Samuel-Fiddis** [comment]
Will this change allow for the basic auth credentials to be passed to the confluent schema registry? i.e. will this PR solve the problem detailed here: https://stackoverflow.com/questions/66096859/failed-to-connect-to-confluent-platform-schema-registry-apache-flink-sql-confl?newreg=e47e09684e404f4f822e417e74016268

**Comment by qingdao81** [comment]
Good question - I just had a closer look into the confluent serializer code. Confluent has their own implementation of a `RestService` which takes care about http requests to the schema registry. The changes I did so far will only make sure that the ssl properties are forwarded to the `SslFactory` which is used by the `RestService`. I can add another change that auth properties are forwarded to the `RestService` as well. I can not really tell you if this then solves the issue from the stackoverflow thread which you mentioned, but in theory it should.

But I also don't really know how that works in combination with the TableEnvironment. For the StreamingEnvironment it was not possible to provide any config properties (except the schema registry url) to the (de)serializer. But in your example you  already provide a bunch of properties.

**Comment by Samuel-Fiddis** [comment]
I've updated and compiled the source code appending the following to the CachedSchemaCoderProvider class with the SchemaCoder get() method:

import io.confluent.kafka.schemaregistry.client.SchemaRegistryClientConfig;
import io.confluent.kafka.schemaregistry.client.security.basicauth.UserInfoCredentialProvider;

@Override
public SchemaCoder get() {
  RestService restService = new RestService(url);
  if(registryConfigs != null && Objects.equals(registryConfigs.get(SchemaRegistryClientConfig.BASIC_AUTH_CREDENTIALS_SOURCE), "USER_INFO")) {
                  UserInfoCredentialProvider provider = new UserInfoCredentialProvider();
                  provider.configure(registryConfigs);
                  restService.setBasicAuthCredentialProvider(provider);
  }
}
....

But it doesn't look like the basic auth properties are passed from the Table API through to the registryConfigs map. When I've hard coded the properties into the registryConfigs everything works fine and the confluent schema registry works with basic auth access.

**Comment by qingdao81** [comment]
I added another commit which forwards the `registryConfig` to the `RestService`:
https://github.com/apache/flink/pull/15808/commits/070930ab00fecf6d07c01f3d02f4894dd8625a3a#diff-5f30e0f1719a976a878d0d07e9e9d47fac7f3a5d6daa7631349439c23f5320ffR72

The `RestService` uses a `BasicAuthCredentialProvider`:

```
public void configure(Map<String, ?> configs) {
        String basicCredentialsSource = (String)configs.get("basic.auth.credentials.source");
        String bearerCredentialsSource = (String)configs.get("bearer.auth.credentials.source");
        if (isNonEmpty(basicCredentialsSource) && isNonEmpty(bearerCredentialsSource)) {
            throw new ConfigException(String.format("Only one of '%s' and '%s' may be specified", "basic.auth.credentials.source", "bearer.auth.credentials.source"));
        } else {
            if (isNonEmpty(basicCredentialsSource)) {
                BasicAuthCredentialProvider basicAuthCredentialProvider = BasicAuthCredentialProviderFactory.getBasicAuthCredentialProvider(basicCredentialsSource, configs);
                this.setBasicAuthCredentialProvider(basicAuthCredentialProvider);
            } else if (isNonEmpty(bearerCredentialsSource)) {
                BearerAuthCredentialProvider bearerAuthCredentialProvider = BearerAuthCredentialProviderFactory.getBearerAuthCredentialProvider(bearerCredentialsSource, configs);
                this.setBearerAuthCredentialProvider(bearerAuthCredentialProvider);
            }

            String proxyHost = (String)configs.get("proxy.host");
            Object proxyPortVal = configs.get("proxy.port");
            Integer proxyPort = proxyPortVal instanceof String ? Integer.valueOf((String)proxyPortVal) : (Integer)proxyPortVal;
            if (isValidProxyConfig(proxyHost, proxyPort)) {
                this.setProxy(proxyHost, proxyPort);
            }

        }
    }
```

That should hopefully do the same as your change.

**Comment by AHeise** [comment]
Hi @qingdao81 , thanks for your contribution. I have done a quick review round and it looks good already. Could you please resolve the compile issue (probably with your latest commit).

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProvider.java
Comment (qingdao81) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProvider.java:71]: Yes indeed - good point I will revert this change. Thanks!

**Comment by qingdao81** [comment]
@tao-xiao-sn You are right the properties are not forwarded to the SQL Factory. I will update this code as well.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java:66]: dots are for hierarchies not for word separation, rewrite to `schema-registry.basic-auth.credentials-source`

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java:72]: rewrite to `schema-registry.basic-auth.user-info`

File: flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchemaTest.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchemaTest.java:34]: Tests should always be well chosen. I have feeling that this test is not very useful and can be dropped.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:100]: avoid code duplication and simplify `map with additional schema registry configs (for example for SSL properties).`

File: flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroSerializationSchemaTest.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroSerializationSchemaTest.java:34]: Remove as well?

File: flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactoryTest.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactoryTest.java:144]: Reduce code duplication and move shared parts with the test above into a helper method.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java
Comment (tao-xiao-sn) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:181]: i dont think this is sufficient, for example schema registry supports bearer token too
https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClientConfig.java#L34

I think a better solution is to pass all properties that start with `schema-registry` prefix to schema registry client so that we dont need be coupled with the implementation of schema registry. this is similar to how table api pass kafka properties to kafka

File: flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProviderTest.java
Comment (twalthr) [inline flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProviderTest.java:56]: is it possible to also check what has been set?

**Review by twalthr — COMMENTED** [review]
Some feedback from my side as well.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java
Comment (twalthr) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:181]: +1 to this, we should not come up with our own properties because this means more maintenance overhead in the long-term when new properties are introduced. A prefix should be sufficient.

Comment (qingdao81) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:181]: Yeah I was thinking about this as well but I wasn't sure how that works with the [RegistryAvroOptions](https://github.com/apache/flink/blob/master/flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java)

To my understanding I have to add one `ConfigOption` for each property?

Comment (twalthr) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:181]: You are right, this is a shortcoming in the current FormatFactory design. In dynamic table factories you have access to the plain `Map<String, String>` for excluding prefixes but not in format factories. We should fix this in a follow up issue. I will create a JIRA ticket for it.

Until then let's do hard coded properties but already in prefixed format. Then we can replace those explicit options with wildcard options in the future.

Comment (twalthr) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactory.java:181]: I created https://issues.apache.org/jira/browse/FLINK-22878. But this might take a while because it would affect many components in Flink to solve it properly.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java
Comment (qingdao81) [reply flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroOptions.java:66]: These properties are forwarded to the confluent `RestService` which expects a `.` notation.

[SchemaRegistryClientConfig](https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClientConfig.java#L25)

That is even true for `schema-registry`. I think for the 2 already existing properties `schema-registry.url` and `schema-registry.subject` this doesn't matter because they are treated in a different way. (they are parsed and then used for confluent api calls)

To my understanding for the additional properties it will only work if we forward for example `schema.registry.basic.auth.credentials.source` to the confluent api. If we use the grouping notation the properties will not match anymore and will be ignored.

If Flink should work with the dot and slash separation for hierarchies I had to transform the input back to pure dot notation at some point before delegating it to the confluent api.

At least this is my understanding but maybe I'm missing something?

**Comment by twalthr** [comment]
@qingdao81 please let me know when you can update the PR. I'm happy to review it again. I think a lot of users could benefit from this.

**Comment by qingdao81** [comment]
@twalthr I was out the last couple of days but I should have time to continue tomorrow or latest on Wednesday.

**Comment by qingdao81** [comment]
@twalthr I updated the MR. I debugged the code and noticed that the `CachedSchemaRegistryClient` expects all properties in a dot notation and additional ssl related properties in the format: `schema.registry.ssl.xxx`
and auth related properties in the format without the `schema.registry` prefix: `basic.auth.user.info`

I still followed your suggestion to define config groups with the combination of dash / dot notation and transformed the Flink specific format to the Confluent specific format in `CachedSchemaCoderProvider#formatRegistryConfigs`.

I removed the unit tests which were not really meaningful and added a some tests for basic and bearer auth.

Happy to get your feedback!

**Review by twalthr — APPROVED** [review]
Thanks for the update @qingdao81. I forgot to delete my comment about the property format . If we simply forward the plain properties (without prefix) in the future, we can use the original keys with dot syntax. I will fix this while merging. Otherwise the changes look good to me. I will also investigate whether we can merge this to 1.13 as well to unblock users.

**Comment by qingdao81** [comment]
@twalthr Sounds good to me. Let me know in case you need any further support.

**Comment by twalthr** [comment]
I have solved a couple of infrastructure problems in the meantime (support for prefix map syntax, support for fallback keys in the FactoryUtil). I pushed the final commits to this PR. The changes should be in a mergeable state now.

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java
Comment (Airblader) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:95]: nit (and same below):

```suggestion
     * Creates {@link ConfluentRegistryAvroDeserializationSchema} that produces {@link
     * GenericRecord} using the provided reader schema and looks up the writer schema in the Confluent Schema
     * Registry.
```

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java
Comment (Airblader) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:115]: Why is the default documented here? How would I pass an `int` that uses the default?

File: flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java
Comment (Airblader) [inline flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:181]: Same as above

**Review by Airblader — APPROVED** [review]
I just have a couple of nit comments on the docs, but LGTM overall and +1.

**Comment by clouddra** [comment]
Are the new options applied to `DebeziumAvroFormatFactory` as well?

**Comment by twalthr** [comment]
@clouddra no they are not :( Seems we forgot to propagate them. I will open a follow up issue.

**Pull request closed without merging**
